### PR TITLE
Reduce project forks on solution open

### PR DESCRIPTION
**Pull request by ToddGrun**
This doesn't have a huge impact on solution open time, so I'm on the fence about whether this is worth the churn. I *think* the new code is better, but I'm flexible about whether to pursue this.

*** CPU from forks (before) ***
image

*** CPU from forks (after) ***
image

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs [hunk 1]
```diff
@@ -810,10 +813,21 @@ static bool IsUnchanged(DocumentState oldDocument, SyntaxNode root)
...
+        ImmutableArray<(DocumentId documentId, DocumentState documentState)> documentIdsAndStates, bool forceEvenIfTreesWouldDiffer)
     {
-        return UpdateDocumentState(
-            this.SolutionState.WithDocumentContentsFrom(documentId, documentState, forceEvenIfTreesWouldDiffer), documentId);
+        return WithDocumentContents(
+            documentIdsAndStates,
+            isUnchanged: static (oldDocumentState, documentState, forceEvenIfTreesWouldDiffer) =>
+            {
+                if (oldDocumentState == documentState)
+                    return true;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs:823]: could remove this if you want.  the below check will result in the same, with virtually the same cost.  but i'm fine with this.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 2]
```diff
@@ -284,90 +284,109 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
...
-            // doc as we'll be done once we update our contents to theirs.  Note: GetFirstRelatedDocumentId will also
-            // not search the project that addedDocumentId came from.  So this will help ensure we don't repeatedly add
-            // documents to a project, then look for related docs *within that project*, forcing the file-path map in it
-            // to be recreated for each document.
-            var relatedDocumentId = solution.GetFirstRelatedDocumentId(addedDocumentId, relatedProjectIdHint);
+            foreach (var addedDocumentId in addedDocumentIds)
+            {
+                // Ensure we don't search in addedDocumentId's project for the related document
+                if (addedDocumentId.ProjectId == relatedProjectIdHint)
+                    relatedProjectIdHint = null;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:329]: is this change necessary?  (i'm guessing yes.  does GetFirstRelatedDocumentId basically assert to make sure the caller isn't doing something dumb?)

**Review by CyrusNajmabadi — APPROVED** [review]
i'm on the fence here.  there's enough complexity that i worry about some subtle regression.  but i don't mind taking this if you want it.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 3]
```diff
@@ -284,90 +284,109 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
...
-                    newSolution = UpdateExistingDocumentsToChangedDocumentContents(newSolution, changedDocumentId, seenChangedDocuments);
+                // Now do the same for all added and changed documents in a project.
+                addedDocumentIds.AddRange(projectChanges.GetAddedDocuments());
+                changedDocumentIds.AddRange(projectChanges.GetChangedDocuments());
             }
 
+            newSolution = UpdateAddedDocumentToExistingContentsInSolution(newSolution, addedDocumentIds);
+
+            // now, for any changed document, ensure we go and make all links to it have the same text/tree.
+            newSolution = UpdateExistingDocumentsToChangedDocumentContents(newSolution, changedDocumentIds);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:314]: i like this change.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 4]
```diff
@@ -284,90 +284,109 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
...
+            return solution.WithDocumentContentsFrom(relatedDocumentIdsAndStates.ToImmutableAndClear(), forceEvenIfTreesWouldDiffer: false);
         }
 
-        static Solution UpdateExistingDocumentsToChangedDocumentContents(Solution solution, DocumentId changedDocumentId, HashSet<DocumentId> processedDocuments)
+        static Solution UpdateExistingDocumentsToChangedDocumentContents(Solution solution, HashSet<DocumentId> changedDocumentIds)
         {
             // Changing a document in a linked-doc-chain will end up producing N changed documents.  We only want to
             // process that chain once.
-            if (processedDocuments.Add(changedDocumentId))
+            using var _ = PooledDictionary<DocumentId, DocumentState>.GetInstance(out var documentIdsToUpdate);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:364]: can this happen?  do we have any case where we'd see a doc-id twice for changed docs.

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 5]
```diff
@@ -284,90 +284,109 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
...
+                        documentIdsToUpdate[relatedDocumentId] = changedDocument.DocumentState;
+                    }
                 }
             }
 
-            return solution;
+            if (documentIdsToUpdate.Count == 0)
+                return solution;
+
+            var toProcessRelatedDocumentIds = documentIdsToUpdate.SelectAsArray(kvp => (kvp.Key, kvp.Value));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:387]: i can't quite reason about everything here.  do we really ever need a map?  couldn't an array work for this?  

also, toProcessRelatedDocumentIds  seems like the wrong name, since it's more than just the id.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:387]:
```suggestion
            var toProcessRelatedDocumentIds = documentIdsToUpdate.SelectAsArray(static kvp => (kvp.Key, kvp.Value));
```

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs [hunk 6]
```diff
@@ -1194,8 +1213,6 @@ protected internal void OnAnalyzerConfigDocumentTextLoaderChanged(DocumentId doc
                         // Have the linked documents point *into* the same instance data that the initial document
                         // points at.  This way things like tree data can be shared across docs.
 
-                        var options = oldSolution.Services.GetRequiredService<IWorkspaceConfigurationService>().Options;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Workspace.cs:1197]: def feel free to file another pr with just this change :)

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs (hunk 2 above)
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs:329]: Exactly. First line in SolutionState.GetFirstRelatedDocumentId:

Contract.ThrowIfTrue(documentId.ProjectId == relatedProjectIdHint);

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs (hunk 4 above)
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs:364]: I was being careful, but it seems like two docs could change that have the same related document, no?

File: src/Workspaces/Core/Portable/Workspace/Workspace.cs (hunk 5 above)
Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Workspace.cs:387]: As I mentioned in the other comment, it seems like it could need a map. I'll change the name.

**Comment by ToddGrun** [comment]
I'm a little more inclined to move forward since you didn't see anything blatantly wrong. Feels nice to get rid of the last big forking issue on solution open.

---
In reply to: [2097221328](https://github.com/dotnet/roslyn/pull/73839#issuecomment-2097221328) [](http://example.com/codeflow?ancestors=2097221328)
